test/e2e_env/universal/gateway/gateway.go

```diff
@@ -0,0 +1,267 @@
+		})
+	})
+
+	Context("when mTLS is enabled", func() {
+		It("should proxy simple HTTP requests", func() {
+			proxySimpleRequests(env.Cluster, "universal",
+				GatewayAddressPort("gateway-proxy-mtls", mtlsGatewayPort), "example-mtls.kuma.io")
+		})
+
+		// In mTLS mode, only the presence of TrafficPermission rules allow services to receive
+		// traffic, so removing the permission should cause requests to fail. We use this to
+		// prove that mTLS is enabled
+		It("should fail without TrafficPermission", func() {
+			proxyRequestsWithMissingPermission(env.Cluster, mtlsMesh,
+				GatewayAddressPort("gateway-proxy-mtls", mtlsGatewayPort), "example-mtls.kuma.io")
+		})
+	})
+
+	Context("when targeting an external service", func() {
+		BeforeAll(func() {
+			// The suite-level JustBeforeEach adds a default route to the mesh echo server.
+			// Add new route to the external echo server.
+			Expect(
+				env.Cluster.GetKumactlOptions().KumactlApplyFromString(fmt.Sprintf(`
+type: MeshGatewayRoute
+mesh: %s
+name: external-routes
+selectors:
+- match:
+    kuma.io/service: edge-gateway
+conf:
+  http:
+    rules:
+    - matches:
+      - path:
+          match: PREFIX
+          value: /external
+      backends:
+      - destination:
+          kuma.io/service: external-echo
+`, mesh)),
+			).To(Succeed())
+
+			Expect(
+				env.Cluster.GetKumactlOptions().KumactlApplyFromString(fmt.Sprintf(`
+type: ExternalService
+mesh: %s
+name: external-service
+tags:
+  kuma.io/service: external-echo
+networking:
+  address: "%s"
+`, mesh, net.JoinHostPort(env.Cluster.GetApp("external-echo-gateway").GetIP(), "8080"))),
+			).To(Succeed())
+		})
+
+		It("should proxy simple HTTP requests", func() {
+			proxySimpleRequests(env.Cluster, "external-echo",
+				GatewayAddressPort("gateway-proxy", gatewayPort), "example.kuma.io",
+				client.WithPathPrefix("/external"))
+		})
+	})
+	Context("applying ProxyTemplate", func() {
+		It("shouldn't error with gateway-proxy import", func() {
+			Expect(
+				env.Cluster.GetKumactlOptions().KumactlApplyFromString(fmt.Sprintf(`
+type: ProxyTemplate
+mesh: %s
+name: edge-gateway
+selectors:
+  - match:
+      kuma.io/service: edge-gateway
+conf:
+  imports:
+    - gateway-proxy
+  modifications:
+    - cluster:
+        operation: add
+        value: |
+          name: test-cluster
+          connectTimeout: 5s
+          type: STATIC
+`, mesh)),
+			).To(Succeed())
+		})
+	})
+
+	Context("when a rate limit is configured", func() {
+		BeforeAll(func() {
+			Expect(
+				env.Cluster.GetKumactlOptions().KumactlApplyFromString(fmt.Sprintf(`
+type: RateLimit
+mesh: %s
+name: echo-rate-limit
+sources:
+- match:
+    kuma.io/service: edge-gateway
+destinations:
+- match:
+    kuma.io/service: echo-service
+conf:
+  http:
+    requests: 5
+    interval: 10s
+`, mesh)),
+			).To(Succeed())
+		})
+
+		It("should be rate limited", func() {
+			proxyRequestsWithRateLimit(env.Cluster,
```

Comment:
could we inline this? It's only used once

Reply:
inlined and rewritten with `CollectFailure`.